components/raftstore/src/store/peer.rs

```diff
@@ -4387,6 +4400,38 @@ where
             }
         }
     }
+
+    pub fn unsafe_recovery_wait_apply<T: Transport>(
+        &mut self,
+        ctx: &PollContext<EK, ER, T>,
+        counter: Arc<AtomicUsize>,
+    ) {
+        self.unsafe_recovery_state = Some(UnsafeRecoveryState::WaitApply {
+            target_index: self.raft_group.store().commit_index(),
+            task_counter: counter,
+        });
+        self.unsafe_recovery_maybe_finish_wait_apply(ctx, /*force=*/ false);
+    }
+
+    pub fn unsafe_recovery_maybe_finish_wait_apply<T: Transport>(
+        &mut self,
+        ctx: &PollContext<EK, ER, T>,
+        force: bool,
+    ) {
+        if let Some(unsafe_recovery_state) = self.unsafe_recovery_state.as_ref() {
+            let UnsafeRecoveryState::WaitApply {
+                target_index,
+                task_counter,
+            } = unsafe_recovery_state;
+            if self.raft_group.raft.raft_log.applied >= *target_index || force {
+                if task_counter.fetch_sub(1, Ordering::Relaxed) == 1 {
+                    store_heartbeat_pd(ctx, /*send_detailed_report=*/ true);
```

Comment:
Why is a store heartbeat sent from a peer? What's the point of abstract then? 🤣

Reply:
Better than sending to PD directly.

Reply:
alright, let's revert the change 🤣

Reply:
Done